### ISSUE TITLE
Types Discussion: how to change Comparable and Equatable behaviours?

### ISSUE THREAD

**Issue by TheMartianLife** [YarnSpinnerTool/YarnSpinner#167]
There have been some comments about how fallback behaviour should work in comparing incompatible or NULL values (see #151, #157, others). The addition of any of possible fixes come with their own issues but our key options are likely:

* Officiate and enforce a stronger type system. This is the most work but arguably the nicest from a purist technical perspective.
* Throwing errors. Intelligent ones. This makes people fix their code which (let's be honest) probably shouldn't be relying on comparison edge case behaviours anyway.
* Doing better inference logic to make falling through less likely. e.g. if the two value types are not the same, try to convert one (right now all compared values fallback to String comparison). This arguably will cause the same problem we have, in a different way, but the "truth table" complexity of a language with 4 types ain't that  bad.

However, each of these will likely break at least some existing code.

**Comment by McJones** [comment]
My vote (for whatever little impact that has...) is for a proper type system.

Could also add in some sort of ~= operator (totally just making up the operator there) which would do a type coercion before of the r value to the l value's type before checking. This would allow for times where you REALLY REALLY REALLY need to check in a coerced fashion. This doesn't resolve the "hey why is null less than 5" sort of issues that might occur but I think it offers the best of both worlds.

**Comment by unknowndevice** [comment]
I'm for a stronger type system. In the short term some kind of checker or explict error would be helpful. As has been mentioned here already more often then not this is happening by accident and not as deliberate request from a user.

In our own team nearly all the dialogue in our game is being written by non-technical user. They know all the features of yarn but don't know how to debug subtle errors like this. Having a explict fail would remove the ambiguity for this level of user.

**Comment by McJones** [comment]
Based on #198 it looks like the time is ripe to open this discussion back up and get an agreement on how this all works for hopefully adding into the 1.1.x or 1.2.x release of Yarn Spinner.
Some of this discussion also impacts how all variables and handled but the main focus of this comment is on comparison.
I see this as having a bunch of different issues all tied together so I will try and describe the issues and the potential solutions they have:

1. How do null comparisons get handled?
2. How do comparisons across types get handled?
3. How do undeclared variables get handled?

One way to handle all three is to have a Comparable function call that Yarn calls out to and let each user provide their own implementation. This has the main advantage that everyone can make their own decision as to what `null < 5` means.
However we'd still have to provide a default version of this and I doubt many people would ever bother providing their own. If someone **truly** hated the comparison system that much they could build their own version of the library instead and use that.
Personally I am against having this approach.

### 1.
The first options is that `null` comparisons should always fail unless both values are `null`.
The concern around this is with situations like `null < 5` and `null > 5` which while they should both evaluate to false it might look a little weird to people who don't get the concept of `null`.

Another option is for `null` to unwrap into whatever the default value is for the value being compared.
So for example assuming that *String*’s default to `””` and *Number*’s default to `0`, `$someNullValue == "hello"` then the comparison will become `"" == "hello"`, in the case of `$someNullValue == 0` it will become `0 == 0`.
The advantage of this is it handles any situation automatically, the disadvantage is we still get weird behaviour like `null` is larger than `-1` which doesn't really make sense.

We can also take a more extreme version of this similar to what [Swift does](https://developer.apple.com/documentation/swift/optional) and ban directly manipulating `null`.
If a null is encountered at runtime then we throw an error.
Then we push the work back onto the user to handle these situations by ensuring they don’t null out any variables before their time.
This version would mean that directly setting variables to `null` should never happen outside of error or you agreeing that you are permanently done with a variable.

### 2.
With **2.** it is a lot simpler: either we allow comparisons across types or we don't.

If we allow comparisons across types we have to work out how we want type cohersion to work, we will need to pick a model and stick with it.
Probably the simplest way is to use C#'s built in type casting and if it fails we fall back to handling the value as a `null` (based on **1.**).
This means that something like `"5" > 2` will evaluate to true but `"hi" > 2` will becomes `null > 2` and likely evaluate to false.

The other question is around which side of the comparison becomes the type to compare against.
In the above examples the l-value was converted into the r-value’s type, but is this always the desired behaviour?
I think it makes sense for literals to not be converted but what do we do in the situations like `$someNumber < $someString`?

The other option is to ban comparison across types, this could be done either as a check by the compiler (which should be doable) or at runtime.
I would prefer if we pick this option and that we go with checking done at compile time.

The real question for this is how often are people actually comparing across types?
Because I haven’t seen a whole lot of this happening or too many issues being raised around it.
If it is very low it might be best to ban it and when people want to do it we tell them they’ve got to do the conversion themselves.

### 3.
For **3.** while it isn't really part of this issue it does tie in and impact the others so it might as well be addressed.
Undeclared variables are currently set to `null`, while we can keep this, is this behaviour still desirable?

First approach is that we don't allow using undeclared variables.
This can then be tracked by the compiler (I think) and can ensure that a variable is set before it is used or throw an error.

The second option is we put some sort of new section into Yarn Spinner that presents a box in Unity where the user can pick their defaults for each type.
So they can choose that all *String* variables will default to `“DEFAULT TEXT”` if not otherwise set, for example.
By default these would all be set to `null` with whatever consequences that has.

### My opinion
My personal choice is that we disallow `null` checks, ban cross type comparisons, and allow the users to set their defaults for undeclared variables.
This is also sorta the extreme position but I think it best solves the problems here by pushing good behaviour as the default.
It forces you to correctly initialise variables, to think through your comparisons, but also offers a way to prototype without getting in the way of having to go “oh wait I forgot to declare $someVar” before using it.

Thoughts?

**Comment by McJones** [comment]
oops... why is close and comment the first option shown...

**Comment by TheMartianLife** [comment]
Going to throw out some weird options here, for the sake of discussion:

I think if our debate is about whether people will understand more nuanced behaviour then maybe this is almost an argument for == versus === operators, as a sort of “fuzzy comparison which people who don’t know/car about the different default to” and “proper exact comparison that will pull you up on your being lazy”. But then that’s about the same as overriding  the == implementation. 

Another slightly weird option is also to have variable-level constraints like:
thing = 1
(thing is now a soft int or ‘any’ type under the hood)
thing.allowsConvertingComparison = false
(thing is now a strict int under the hood)

Either way I think null has a place, but I’d like to hear a) some views from non-programmers who don’t know what it is about what they’d assume it does/is for and b) some consideration for whether better debugging/tracing tools would negate some issues with it. Like logs warnings when you run a check against null and it resolves false, or whatever. 

Or even there’s like a derivative one of Python’s approaches which could be considered, where null must be explicitly checked against. Like if you if thing == otherThing and thing is null then it errors but if you check if thing is null then it passes as a check?

Just some thoughts 🙃

**Comment by McJones** [comment]
Personally I am totally against having a `==` and `===` operators, in my experience they only seem to add more confusion and weird rules of thumbs like *"make sure to always use === when using Strings"* etc etc without actually providing any insight why.
I see that as just pushing the type checking onto the person, which is what we basically have now.

Dynamically changing the type checking is an interesting idea but considering that it requires a person to do it, wouldn't adding in casting functions make more sense and work like how C# already works?
Making it toggle on and off at a variable level just feels like a time bomb of forgetting a toggle written long ago.
Maybe it's just the old school Obj-C programmer in me but I'm getting *"you forget the* `[object release]` *message didn't you?"* flashbacks with this idea.
I do like the idea of making this something that can be toggled on and off system wide though.
So during development you don't care and let it munge stuff together how it sees fit but in release and testing you force the type checking.

Not entirely sure what you mean by the Python approach, doesn't Python just treat None as its own type and prevents all comparisons (or actions of any kind) with it?
We essentially have that now, you can already do:

```
<<if $someVar eq Null>>
// log it or something, I don't know...
<<else>>
// do whatever you wanted to do originally
<<endif>>
```

Making people do null checks feels the clunkiest way possible.
I believe we should be trying to minimise encountering null, not making it explicit.

**Comment by naNuke** [comment]
I have fell into the trap when I was trying yarn for first time as well and shortly figured out that my script doesnt do things I want it to when <<if $undefined eq 2>> failed. 

Ideally, for people who can't code at all but would like to use Yarn as part of a team or by themselves might fall into this as well and get stuck for a while. 

So personally I would vote for disallowing nulls altogether and enforce default values. It is not really that hard for people to declare their variables if they are aware of that kind of rule by reading the documentation.

My thoughts are that while I am all for enforcing default variables, it feels kind of weird to not provide one place to declare those, imagine a scenario where I have bunch of related nodes in a yarn file but not a common starting node. 

That said, maybe it would be worth to consider a common place for default variables instead of making people set them possibly in multiple nodes and think about if they havent forget somewhere.

The very top of .yarn file above first node maybe?

```
$variable1 0  
$variable2 "string"  
$gold 1000
$haveMetShadyGuy false

title: Start
... nodes declared here
```

This would be both straightforward and easy to use in my opinion.

Additionaly variable storage could fetch those variables right away, Yarn Compiler could provide those in inspector and warn users if code wants to use variable that has not been set etc etc.

TLDR: enforce default, disallow null, prevent comparisions between types and provide one place to set default variables in the yarn file.

**Comment by desplesda** [comment]
With the static typing support now preset in v2.0.0-beta2, this issue no longer applies, so I'm closing it! Thanks to everyone who contributed to the discussion, which led directly to the current type system.
